arch/xtensa/src/esp32s2/esp32s2_spiram.c

```diff
@@ -0,0 +1,450 @@
+                  uint32_t *out_start_vaddr)
+{
+  /* For now, this function should only run when virtual address is enough
+   * Decide these logics when there's a real PSRAM with larger size
+   */
+
+  uint32_t map_length = end_paddr - start_paddr;
+
+  if (map_length > SOC_EXTRAM_DATA_SIZE)
+    {
+      /* Decide these logics when there's a real PSRAM with larger size */
+
+      merr("PSRAM physical size is too large, not support mapping it yet!");
+      return -ENOMEM;
+    }
+
+  /* should be MMU page aligned */
+
+  assert((start_paddr % MMU_PAGE_SIZE) == 0);
+
+  uint32_t start_vaddr = DPORT_CACHE_ADDRESS_LOW;
+  uint32_t end_vaddr = start_vaddr + map_length;
+  uint32_t cache_bus_mask = 0;
+
+  cache_bus_mask |= (end_vaddr > 0) ? EXTMEM_PRO_DCACHE_MASK_DPORT : 0;
+  cache_bus_mask |= (end_vaddr >= DPORT_ADDRESS_HIGH) ?
+                    EXTMEM_PRO_DCACHE_MASK_DRAM1 : 0;
+  cache_bus_mask |= (end_vaddr >= DRAM1_ADDRESS_HIGH) ?
+                    EXTMEM_PRO_DCACHE_MASK_DRAM0 : 0;
+
+  assert(end_vaddr <= DRAM0_CACHE_ADDRESS_HIGH);
+
+  minfo("start_paddr is %x, map_length is %xB, %d pages",
+        start_paddr, map_length, BYTES_TO_MMU_PAGE(map_length));
+
+  /* No need to disable cache, this file is put in Internal RAM */
+
+  cache_dbus_mmu_set(MMU_ACCESS_SPIRAM, start_vaddr, start_paddr, 64,
+                     BYTES_TO_MMU_PAGE(map_length), 0);
+
+  REG_CLR_BIT(EXTMEM_PRO_DCACHE_CTRL1_REG, cache_bus_mask);
+
+  *out_start_vaddr = start_vaddr;
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: mmu_map_psram
+ *
+ * Description:
+ *   Initialize the CACHE to use with PSRAM
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void IRAM_ATTR esp_spiram_init_cache(void)
+{
+  int ret;
+  static DRAM_ATTR uint32_t vaddr_start = 0;
+  uint32_t start_page = 0;
+  uint32_t psram_available_size = esp_spiram_get_size();
+
+  /* Map the PSRAM physical range to MMU */
+
+  ret = mmu_map_psram(MMU_PAGE_TO_BYTES(start_page),
+                      MMU_PAGE_TO_BYTES(start_page) +
+                      psram_available_size, &vaddr_start);
+  if (ret < 0)
+    {
+      merr("MMU PSRAM mapping wrong!");
+      abort();
+    }
+
+  /* After mapping, we DON'T care about the PSRAM PHYSICAL
+   * ADDRESSS ANYMORE!
+   */
+
+  g_allocable_vaddr_start = vaddr_start;
+  g_allocable_vaddr_end   = vaddr_start + psram_available_size;
+}
+
+/****************************************************************************
+ * Name: esp_spiram_test
+ *
+ * Description:
+ *   Simple RAM test. Writes a word every 32 bytes. Takes about a second
+ *   to complete for 4MiB. Returns true when RAM seems OK, false when test
+ *   fails. WARNING: Do not run this before the 2nd cpu has been initialized
+ *   (in a two-core system) or after the heap allocator has taken ownership
+ *   of the memory.
```

Comment:
```suggestion
 *   fails. WARNING: Do not run this after the heap allocator has taken ownership
 *   of the memory.
```

Reply:
Why do we need this function at all?  We have the app `ramtest` that does the same thing.

Reply:
I kept it because the user can test the PSRAM before getting the system initialized.